### DISCUSSION TITLE
Table view and grid system

### DISCUSSION THREAD

**Discussion by himself65** [toeverything/blocksuite#717] [General]
Currently, our data structure will or may limit the grid system or database block implementation. Why? Our infra code is a tree structure that each model holds its children. This may affect the code spec of the drag handle (seachBinary part by @fundon), and the way to make the tree into one depth array.

**Comment by himself65** [comment]
# RFC: Row Block

This is a model that will display its children in the row direction.

Pseudo code example:

```tsx
<database-block>
  <row-block>
    <paragraph-block id=1/>
    <paragraph-block id=2/>
  </row-block>
  <row-block>
    <paragraph-block id=3/>
  </row-block>
</database-block>
```

Turned into ???

## Edge Cases

### Drag Handle

binary search will have the fallback if it detects one block inside a row block

**Comment by doodlewind** [comment]
A draft design of the RowBlock:

<img width="580" src="https://user-images.githubusercontent.com/7312949/212530370-908b7b53-7741-469e-af34-b4509493e6af.jpeg">

Based on this design, it's possible that we support GridBlock in a similar way (having the grid container that holds the `columns`, rather than on the database).

<img width="580" alt="image" src="https://user-images.githubusercontent.com/7312949/212530548-bf30aab1-f6b7-4187-8f2e-e15968720a8f.png">

In this case, the outer DatabaseBlock won't have `columns`. This blurs the boundary between DatabaseBlock and GridBlock. In this case, essentially we only need a ContainerBlock that holds either:

* `ContainerBlock[]`, in this case, the parent container is used for layout.
* `ContentBlock[]`, in this case, the parent container is used for column-and-tag-based view.

**Comment by fundon** [comment]
Within DatabaseBlock, there is its own drag-and-drop behavior, such as row, column dragging, etc. As well as it supports only simple primitive types.

In the context of a document, it is like any other block, the whole DatabaseBlock can be dragged freely. But other blocks cannot be dragged into the DatabaseBlock unless the DatabaseBlock supports advanced data types.

If other types of blocks are to be dragged into the DatabaseBlock, a type(block) conversion is required.

**Reply by himself65** [reply]
There is one thing wrong, there will have other blocks inside a database block. And users are freely to drag and drop any block inside and out

**Reply by fundon** [reply]
If we think of a cell as a small document, any type of block can indeed be freely dragged in and out.

But if it is a strongly constrained type, such as number, blocks of other types can be dragged in, but the final result is an empty text.

Also, will database blocks support fast import? For example, CSV files.

**Reply by himself65** [reply]
you can see the example database here https://github.com/toeverything/blocksuite/pull/720/files#diff-ff01aa12c5c4252efc977816dac5fcb587bc9316e76fd98dcce2c586f281117aR84-R153

**Reply by himself65** [reply]
We have two different type: blockColumn and tagColumn

**Reply by himself65** [reply]
For the fast-import you mentioned, of course we will support and it not hard

**Reply by doodlewind** [reply]
`TagColumn` is used for columns containing enum-like tags (each cell in that column has the same tag type but with different tag values), and the cells with the `BlockColumn` type can be used to contain block(s).

**Reply by himself65** [reply]
I think tag column is actually the normal type in other products like airtbale, which you can have like phone tag, email tag, number tag, string tag(just a normal string) 

<img width="398" alt="image" src="https://user-images.githubusercontent.com/14026360/212590149-d311bb65-2cd8-489b-843b-cfe7ba4a9898.png">

**Reply by fundon** [reply]
Very cool, looking forward to it.

**Reply by himself65** [reply]
And another difference is we don't hold a cell in the database. it's an attribute(tag) for a block. That makes sense if you drag a row into another database or under the frame block. The attribute won't disappear because it's saved in the top-level of page

**Reply by fundon** [reply]
Interesting.

Another small question is the concept of having a reference block, links a link.

I wouldn't want to drag a block anywhere, I'd just reference it, click it and then jump or expand it.

**Reply by himself65** [reply]
The reference is not in discussion, but I can say that we would have a single Y.Doc to keep all blocks that would be shared in many docs, called `Gallary.`

In our infra, if you want to make a block to be referenced in many docs, first we will put the block into Gallary, then add a reference block(like a pointer)

**Reply by fundon** [reply]
Thanks for the answer. :)

**Reply by himself65** [reply]
![20230116-114449](https://user-images.githubusercontent.com/14026360/212594549-32e97055-31c9-4aec-9cf4-51d7093c3a09.png)
